duration.go

```diff
@@ -22,28 +26,48 @@ func (dur Duration) MarshalText() ([]byte, error) {
 	return []byte(fmt.Sprintf("%02d:%02d:%02d.%03d", h, m, s, ms)), nil
 }
 
-// UnmarshalText implements the encoding.TextUnmarshaler interface.
-func (dur *Duration) UnmarshalText(data []byte) (err error) {
-	parts := strings.SplitN(string(data), ":", 3)
+// DecodeDuration converts a timestamp into a Duration nanoseconds
+func DecodeDuration(data []byte) (*Duration, error) {
+	s := string(data)
+	if len(s) == 0 || strings.ToLower(s) == "undefined" {
+		return nil, ErrDurationEmpty
+	}
+	parts := strings.SplitN(s, ":", 3)
 	if len(parts) != 3 {
-		return fmt.Errorf("invalid duration: %s", data)
+		return nil, fmt.Errorf("invalid duration: %s", data)
 	}
+	var dur = Duration(0)
 	if i := strings.IndexByte(parts[2], '.'); i > 0 {
 		ms, err := strconv.ParseInt(parts[2][i+1:], 10, 32)
 		if err != nil || ms < 0 || ms > 999 {
-			return fmt.Errorf("invalid duration: %s", data)
+			return nil, fmt.Errorf("invalid duration: %s", data)
 		}
 		parts[2] = parts[2][:i]
-		*dur += Duration(ms) * Duration(time.Millisecond)
+		dur += Duration(ms) * Duration(time.Millisecond)
 	}
 	f := Duration(time.Second)
 	for i := 2; i >= 0; i-- {
 		n, err := strconv.ParseInt(parts[i], 10, 32)
 		if err != nil || n < 0 || n > 59 {
-			return fmt.Errorf("invalid duration: %s", data)
+			return nil, fmt.Errorf("invalid duration: %s", data)
 		}
-		*dur += Duration(n) * f
+		dur += Duration(n) * f
 		f *= 60
 	}
+	return &dur, nil
+}
+
+// UnmarshalText implements the encoding.TextUnmarshaler interface.
+func (dur *Duration) UnmarshalText(data []byte) (err error) {
+	// swallow the error since parsing a non-compliant duration will throw an error
+	if d, err := DecodeDuration(data); err != nil {
+		if err == ErrDurationEmpty {
+			*dur = 0
+		} else {
+			return err
+		}
+	} else {
+		*dur = *d
+	}
 	return nil
 }
```

Comment:
you're right. there isn't a reason to have empty error, since we end up catching and returning 0 in the unmarshal.